### ISSUE TITLE
Task and typechain on first run

### ISSUE THREAD

**Issue by sshmaxime** [NomicFoundation/hardhat#1511]
When a task need  the typechain of its project, 1st build/run fails (as typechain is not yet build).

How to tackle that issue ?

**Comment by alcuadrado** [comment]
Hey @sshmaxime,

One workaround for this is running `TS_NODE_TRANSCRIPT_ONLY=1 npx hardhat compile` the first time.

Not idea, but would work.

We are trying to come up with a way to enable typechain in the config automatically, but it's kind of a chicken & egg problem.

**Comment by zgorizzo69** [comment]
I think it is `TS_NODE_TRANSPILE_ONLY=1 npx hardhat compile`

**Comment by sshmaxime** [comment]
@alcuadrado, ideally, #1518 should fix this, no ?

**Comment by alcuadrado** [comment]
Yep :) that's also why I like it so much.

**Comment by paulrberg** [comment]
I've recently upgraded to TypeScript v4.4.2 and the `TS_NODE_TRANSCRIPT_ONLY=1` hack doesn't work anymore:

```
Error: Cannot find module '../../typechain/factories/PRBProxy__factory'
Require stack:
- /Users/paulrberg/projects/prb-proxy/tasks/deploy/prbProxy.ts
- /Users/paulrberg/projects/prb-proxy/tasks/deploy/index.ts
- /Users/paulrberg/projects/prb-proxy/hardhat.config.ts
- /Users/paulrberg/projects/prb-proxy/node_modules/hardhat/internal/core/config/config-loading.js
- /Users/paulrberg/projects/prb-proxy/node_modules/hardhat/internal/cli/cli.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:94:18)
    at Object.<anonymous> (/Users/paulrberg/projects/prb-proxy/tasks/deploy/prbProxy.ts:6:1)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Module.m._compile (/Users/paulrberg/projects/prb-proxy/node_modules/ts-node/src/index.ts:1310:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/paulrberg/projects/prb-proxy/node_modules/ts-node/src/index.ts:1313:12)
    at Module.load (node:internal/modules/cjs/loader:981:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/paulrberg/projects/prb-proxy/tasks/deploy/prbProxy.ts',
    '/Users/paulrberg/projects/prb-proxy/tasks/deploy/index.ts',
    '/Users/paulrberg/projects/prb-proxy/hardhat.config.ts',
    '/Users/paulrberg/projects/prb-proxy/node_modules/hardhat/internal/core/config/config-loading.js',
    '/Users/paulrberg/projects/prb-proxy/node_modules/hardhat/internal/cli/cli.js'
  ]
}
```

I have to manually open `hardhat.config.ts`, comment the import of the tasks which import TypeChain, run the TypeChain task, then uncomment the import.

**Comment by alcuadrado** [comment]
Have you also upgraded ts-node somehow?

**Comment by paulrberg** [comment]
I have .. but weirdly enough I was not able to replicate the issue in my [solidity-template](https://github.com/paulrberg/solidity-template).

The problem only happens in [prb-proxy](https://github.com/paulrberg/prb-proxy). e.g. in CI:

https://github.com/paulrberg/prb-proxy/runs/3560462701?check_suite_focus=true

Can you tell what this is about just looking at the [package.json](https://github.com/paulrberg/prb-proxy/blob/e622aded1ba9eb8bf5fdd556c965a84764f3db8b/package.json)

**Comment by paulrberg** [comment]
Can confirm that this is still happening with hardhat v2.8.4, typechain v7.0.0 and typescript v4.5.5.

**Comment by zgorizzo69** [comment]
tested with hardhat v2.8.4, typechain v7.0.0 and typescript v4.5.5 and running compile with ts node transpile `TS_NODE_TRANSPILE_ONLY=1 npx hardhat compile` is still necessary

**Comment by alcuadrado** [comment]
This should be fixed in v2.11.0. Please reopen the issue if it's still present.
